### ISSUE TITLE
Is this problem because of batch_size?

### ISSUE THREAD

**Issue by gaobaoding** [Microsoft/singleshotpose#83]
Hallo tekin,

I have just reimplement this project with **python3.6 and pytorch1.0.1**, and because of the limited computation of the **GPU(GTX 970m, memory: 3GB)** on my laptop. I changed the **batch_size from 32 to only 2** and trained for a whole night(epoch:700 remains).

after that I test my model, but with results below:
```
-----------------------------------
  tensor to cuda : 0.000402
         predict : 0.004263
get_region_boxes : 0.063616
            eval : 0.009278
           total : 0.077559
-----------------------------------
2019-04-11 13:23:42 Results of ape
2019-04-11 13:23:42    Acc using 5 px 2D Projection = 0.00%
2019-04-11 13:23:42    Acc using 10% threshold - 0.0103 vx 3D Transformation = 0.00%
2019-04-11 13:23:42    Acc using 5 cm 5 degree metric = 0.00%
2019-04-11 13:23:42    Mean 2D pixel error is 2978.765381, Mean vertex error is 2.461958, mean corner error is 409.030334
2019-04-11 13:23:42    Translation error: 2.461311 m, angle error: 143.237230 degree, pixel error:  2978.765628 pix
```

**all the Acc are Zero!!**

I have checked the code, but everything is fine. Do you think it is because of batch_size is too small?? Or still because of the version of python and pytorch.

Looking forward to your reply!

**Comment by btekin** [comment]
Could you try https://github.com/btekin/singleshot6Dpose, for a version that should be compatible with your python and pytorch installations and let me know if this works? There might potentially be something wrong in your re-implementation (e.g. integer division difference between Python 2 & 3 etc.).

Batch size indeed is small, but I think it shouldn't be the main cause for this problem.

**Comment by gaobaoding** [comment]
Actually I work based on https://github.com/btekin/singleshot6Dpose, there's some bugs in the code and after I fixed it, I still got this issue. So I come back and write this open issue. Or do you think my epoch is too small since I reduce the batch_size to 2, maybe 700 is too small? But all results are zero is really strange.

**Comment by btekin** [comment]
What exactly did you change/re-implement/fix? It should in principle work without any changes, at least it is the case for me and some other people working on the project.

**Comment by gaobaoding** [comment]
mostly refer to munti_obj_pose_estimation folder. For example: 
the train_occlusion.txt under multi_obj_pose_estimation/cfg/ is fixed coded path.
and some division problems still exists in multi_obj_pose_estimation folder.
and some dimention problems: e.g. nProposals = int((conf > 0.25).sum().data[0]), I change .data[0] to .data because python raises error that it is non-dimentional tensor.

**Comment by gaobaoding** [comment]
Regardless of the multi tasks. Anyway I will rerun the original code without any modification under https://github.com/btekin/singleshot6Dpose tonight only for single object detection to see if there's still got the same issue. I will update you tomorrow:)
Thanks!

**Comment by gaobaoding** [comment]
hallo tekin,
After testing last night with original code, I still got this problem. All acc are zero. I think the reason is below:
1. The threshold is too small, I didn't got even one expected result during training, so the acc is always 0, and at the result, the best weights will never be updated, either the model weights will never be stored during training.

I have tried to change the threshold for acc2d from 5 to 100, it works, the results are not 0 anymore, but after some steps, the loss become to "nan" or very bigger, I think it is because of underfitted right?

And I would ask for your experience, if it is the same as you train the model, that the three acc are 0 at the beginning during many epochs when you set the threshold to this small. And then after many epochs, the model becomes better and the acc are not 0 anymore?

Thanks!

**Comment by btekin** [comment]
Are you working on the LINEMOD dataset or some other custom dataset that you have? For my experiments on LINEMOD, the accuracy improves in a short time, the accuracy is never zero after a couple of epochs. 

You could also check if the problem stems from the issue in #84. Can you make sure your 2D keypoint predictions are accurate before pnp? Can you check if your mean corner error is reasonable?

**Comment by gaobaoding** [comment]
I working on the LINEMOD dataset. Let me first doing it on a powerful GPU. I'll update you after tried.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by gaobaoding** [comment]
@btekin 
Hi, after change another GPU with 4GB memory and change the batch_size from 2 to 8. The ACC is not zero any more. It is really related with batch_size. I'm still a bit confused with this issue. So do you have any idea why the batch_size cannot be so small just like 2?

**Comment by btekin** [comment]
With very small batch sizes, the learning rate is also set to a value in the implementation. See https://github.com/microsoft/singleshotpose/blob/master/train.py#L386 . Therefore, with small batch sizes you might want to adapt your learning rate as well for better convergence.
